### PR TITLE
Add `abortUnless` analysis

### LINKED ISSUE #873: Support for `void assume_abort_if_not(int cond) {   if(!cond) {abort();} }`
The sv-comp people are really fond of encoding conditions that are needed to later establish correctness into calls to this function. We do not benefit from any such information at all, which we should probably fix for this year's SV-Comp.

~~~C
void assume_abort_if_not(int cond) {
  if(!cond) {abort();}
}
~~~

There are ~3000 instances of this pattern in the non-cwe part of the sv-comp `.c` files, and for some folders (e.g. `weaver`) they seem to have encoded everything into such calls :upside_down_face:

~~~C
  assume_abort_if_not(counter != 0);
  reach_error();
~~~

What I would envision would be a very cheap analysis to determine if a function is such an `assume_abort_if_not_function`:

- Takes one argument of type `int`
- Contains no assigns
- Returns only for non-zero values

If this is the case, I would meet the caller state with the condition that is passed as the argument.

### PR DESCRIPTION
**Pull request by michael-schwarz**
Adds an analysis that detects if control flow only returns if the argument is `non-zero`, and issues a corresponding `assert` event.

TODO:

- [x] Exploit in more analyses
- [x] Test

Closes #873

### CODE REVIEW

**Comment by michael-schwarz** [comment]
There are no other analysis where adding it would make sense (except maybe `vareq`, but that doesn't even exploit `branch` at all currently).
Adding this analysis to the set of default analyses does not cause any regressions.

**Comment by michael-schwarz** [comment]
Unless anyone opposes, I would propose we do a joint sv-comp run of this & #869 to reduce the number of needed runs, as we currently have a lot of things we would need server time for.

File: src/analyses/apron/apronAnalysis.apron.ml [hunk 1]
```diff
@@ -618,6 +618,8 @@ struct
       Priv.enter_multithreaded (Analyses.ask_of_ctx ctx) ctx.global ctx.sideg st
     | Events.Escape escaped ->
       Priv.escape ctx.node (Analyses.ask_of_ctx ctx) ctx.global ctx.sideg st escaped
+    | Assert exp ->
+      branch ctx exp true
```
Comment (sim642) [inline src/analyses/apron/apronAnalysis.apron.ml:622]: `assert_fn` would be more appropriate.

File: src/analyses/base.ml [hunk 2]
```diff
@@ -2810,6 +2810,8 @@ struct
     | Events.AssignSpawnedThread (lval, tid) ->
       (* TODO: is this type right? *)
       set ~ctx (Analyses.ask_of_ctx ctx) ctx.global ctx.local (eval_lv (Analyses.ask_of_ctx ctx) ctx.global ctx.local lval) (Cilfacade.typeOfLval lval) (`Thread (ValueDomain.Threads.singleton tid))
+    | Events.Assert exp ->
+      branch ctx exp true
```
Comment (sim642) [inline src/analyses/base.ml:2814]: `assert_fn` would be more appropriate.

File: src/analyses/abortUnless.ml [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+      match args with
+      | [arg] -> ctx.emit (Events.Assert arg)
+      | _ -> ()
+    );
+    false
+
+  let special ctx (lval: lval option) (f:varinfo) (arglist:exp list) : D.t =
+    let desc = LibraryFunctions.find f in
+    match desc.special arglist with
+    | Abort -> ctx.local
```
Comment (sim642) [inline src/analyses/abortUnless.ml:60]: `Abort` is deadcode anyway, so this doesn't matter at all.

File: src/analyses/abortUnless.ml [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  module C = Lattice.Unit
+
+  let context _ _ = ()
+
+  (* transfer functions *)
+  let assign ctx (lval:lval) (rval:exp) : D.t =
+    false
+
+  let branch ctx (exp:exp) (tv:bool) : D.t =
+    ctx.local
```
Comment (sim642) [inline src/analyses/abortUnless.ml:21]: Any branch? Not just `arg` or `!arg`?

File: src/analyses/abortUnless.ml [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+           | None -> false)
+      | _ ->
+        (* should not happen, ctx.local should always be false in this case *)
+        false
+    else
+      false
+
+  let enter ctx (lval: lval option) (f:fundec) (args:exp list) : (D.t * D.t) list =
+    let candidate = match f.sformals with
+      | [arg] when isIntegralType arg.vtype -> true
```
Comment (sim642) [inline src/analyses/abortUnless.ml:44]: This allows calls with similar signature to be called within `assume_abort_if_not`, including (mutually) recursively. It's not obvious why something bad cannot come from that.

File: src/analyses/abortUnless.ml (hunk 4 above)
Comment (michael-schwarz) [reply src/analyses/abortUnless.ml:21]: Yes, branching does not actually modify the value here, so any branching is fine.

File: src/analyses/abortUnless.ml (hunk 5 above)
Comment (michael-schwarz) [reply src/analyses/abortUnless.ml:44]: Nice catch! It's not the recursion though that was problematic, as we go to `false` in `combine` either way but the fact that the assert happens after evaluation of the transfer function for call, so if variables in `arg` are assigned to, asserting might unsoundly yield bot (as in the case you constructed).

**Comment by michael-schwarz** [comment]
For sv-comp this analysis currently does not seem to pay off, so I would not include it in the config this year (unless I can come up with a more powerful analysis where we would benefit).
I would still propose to merge it, it does not add much code and might be handy.

**Pull request merged**